### ISSUE TITLE
fetch fails if flat=yes, dest=existing-local-directory without trailing slash

### ISSUE THREAD

**Issue by joergrathlev** [ansible/ansible#27312]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
fetch module

##### ANSIBLE VERSION
```
ansible 2.3.1.0
  config file = /Users/[...]/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### OS / ENVIRONMENT
Running from: macOS 10.12.5 (Sierra)
Managing: CentOS 7.2

##### SUMMARY
When using the fetch module, and `dest` is an exisiting local directory, and `dest` is specified without a trailing slash, the file is downloaded into the directory but the checksum check fails after the download.

##### STEPS TO REPRODUCE
```
# Create a test file on the remote system
$ ansible [servername] -m shell -a 'echo "hello world" > /tmp/hello-world'

# Create a directory locally, then try to fetch the file into that directory
$ mkdir local-tmp
$ ansible [servername] -m fetch -a 'src=/tmp/hello-world dest=local-tmp flat=yes'
[servername] | FAILED! => {
    "changed": false, 
    "checksum": null, 
    "dest": "/Users/[...]/local-tmp", 
    "failed": true, 
    "file": "/tmp/hello-world", 
    "md5sum": null, 
    "msg": "checksum mismatch", 
    "remote_checksum": "22596363b3de40b06f981fb85d82312e8c0ed511", 
    "remote_md5sum": null
}

# Note that the file was downloaded correctly
$ cat local-tmp/hello-world 
hello world
```

##### EXPECTED RESULTS
The command should fetch the file, and the checksum check should not fail if the file was downloaded correctly.

(Alternatively, the command should not download the file and report as an error that `dest` already exists.)

##### ACTUAL RESULTS
Checksum mismatch.

##### ADDITIONAL NOTES
After a short look into the code, I believe that instead of checking for a trailing slash, the check should be "trailing slash or existing existing directory" here:
https://github.com/ansible/ansible/blob/ff22528b076fa7ab17e096c73f5d30d370bd0448/lib/ansible/plugins/action/fetch.py#L123

**Comment by abadger** [comment]
We're leaning towards the correct behaviour being that the file should not be downloaded but I haven't gotten a whole lot of feedback yet.  I've added this to the public IRC meeting to get more input.

https://github.com/ansible/community/issues/162#issuecomment-318209405

**Comment by abadger** [comment]
Okay, we talked about this in today's public IRC meeting and decided that failing is the correct choice here as the user may have intended the dest= to be the file and so we'd be silently doing the wrong thing if we made downloading into the directory the correct behaviour.

So we probably need a check inside of the if flat conditional ( https://github.com/ansible/ansible/blob/ff22528b076fa7ab17e096c73f5d30d370bd0448/lib/ansible/plugins/action/fetch.py#L122 ) which checks whether the dest does not end with a slash and the dest is a directory.  If so, error out.
